src/trace/context.ts

```diff
@@ -69,14 +73,66 @@ export function addTraceContextToXray(traceContext: TraceContext) {
     "trace-id": traceContext.traceID,
   };
 
-  captureFunc(xraySubsegmentName, (segment) => {
-    segment.addMetadata(xraySubsegmentKey, val, xraySubsegmentNamespace);
-  });
+  addXrayMetadata(xraySubsegmentKey, val);
 }
 
 export function addStepFunctionContextToXray(context: StepFunctionContext) {
-  captureFunc(xraySubsegmentName, (segment) => {
-    segment.addMetadata(xrayBaggageSubsegmentKey, context, xraySubsegmentNamespace);
+  addXrayMetadata(xrayBaggageSubsegmentKey, context);
+}
+
+export function addXrayMetadata(key: string, metadata: Record<string, any>) {
+  const segment = generateXraySubsegment(key, metadata);
+  if (segment === undefined) {
+    return;
+  }
+  sendXraySubsegment(segment);
+}
+
+export function generateXraySubsegment(key: string, metadata: Record<string, any>) {
+  const header = process.env[xrayTraceEnvVar];
+  if (header === undefined) {
+    logDebug("couldn't read xray trace header from env");
+    return;
+  }
+  const context = parseXrayTraceContextHeader(header);
+  if (context === undefined) {
+    logDebug("couldn't parse xray trace header from env");
+    return;
+  }
+  const time = Date.now();
+
+  return JSON.stringify({
+    id: randomBytes(8).toString("hex"),
+    trace_id: context.xrayTraceID,
+    parent_id: context.xrayParentID,
+    name: xraySubsegmentName,
+    start_time: time,
+    end_time: time,
+    type: "subsegment",
+    metadata: {
+      [xraySubsegmentNamespace]: {
+        [key]: metadata,
+      },
+    },
+  });
+}
+
+export function sendXraySubsegment(segment: string) {
+  const xrayDaemonEnv = process.env[awsXrayDaemonAddressEnvVar];
+  let port = xrayUDPPort;
+  let address = localHost;
+  if (xrayDaemonEnv !== undefined) {
+    const parts = xrayDaemonEnv.split(":");
+    if (parts.length > 1) {
```

Comment:
Is there any sense in continuing if we can't parse the xrayDaemonEnv? It feels like we should mock that in tests, and fail + debug log if we can't find it in production

Reply:
Yeah, you're right, it probably makes sense to bail in that case. I was thinking I'd just copy what the xray-sdk does, which is to have the defaults. But I guess if it's not set, then we probably aren't in a lambda environment

Reply:
Okay, I've implemented this feedback